frontend/src/host_orchestrator/orchestrator/userartifacts.go

```diff
@@ -79,3 +84,32 @@ func (m *UserArtifactsManagerImpl) ListDirs() (*apiv1.ListUploadDirectoriesRespo
 	}
 	return &apiv1.ListUploadDirectoriesResponse{Items: dirs}, nil
 }
+
+func (m *UserArtifactsManagerImpl) GetFullPath(dir, filename string) string {
+	return m.RootDir + "/" + dir + "/" + filename
+}
+
+func (m *UserArtifactsManagerImpl) CreateUpdateArtifact(dir, filename string, src io.Reader) error {
+	dir = m.RootDir + "/" + dir
+	if ok, err := fileExist(dir); err != nil {
+		return err
+	} else if !ok {
+		return operator.NewBadRequestError("upload directory %q does not exist", err)
+	}
+	dst, err := ioutil.TempFile("", "cutfArtifact")
```

Comment:
why copy to a tmp file and then move to final dest? Wouldn't it be better/faster to just copy directly to final dest?

Reply:
performance wise should be quite similar, used a temp file just to avoid having some corrupted file if copying fails for any reason.

Reply:
wouldn't it be the same if the rename fails? Rename is basically a copy and delete if done across file systems in linux IIRC (and tmp is commonly an in-memory file system different from the root disk)

Reply:
good point, didn't considered moving across file system, so I created the temp file in the same directory so rename does not result on copy/delete.

continued using the tmp file idea to allow concurrent updates for the same filename.